Add istio support to k8s operator

Owen-CH-Leung opened the pull request:
This PR extends the contributions of @joshuayeung, advancing the effort to integrate Istio support into the `KubernetesPodOperator`. Previous PR [here](https://github.com/apache/airflow/pull/26739)

joshuayeung committed:
Added IstioKubernetesPodOperator and IstioPodManager

joshuayeung committed:
Added IstioKubernetesPodOperator and IstioPodManager

joshuayeung committed:
Merged into KubernetesPodOperator(istio_enabled=True)

joshuayeung committed:
Merged into KubernetesPodOperator(istio_enabled=True)

Owen-CH-Leung committed:
Add support to istio

Continuing from joshua's work to integrate istio support into KubernetesPodOperator

Owen-CH-Leung committed:
Add support to istio

Continuing from joshua's work to integrate istio support into KubernetesPodOperator

Owen-CH-Leung committed:
Add istio into spelling word

Fix failing spell check

Owen-CH-Leung committed:
Add istio into spelling word

Fix failing spell check

Owen-CH-Leung committed:
Merge branch 'main' into add-istio-support-to-k8sOperator

Owen-CH-Leung committed:
Merge branch 'main' into add-istio-support-to-k8sOperator

Owen-CH-Leung committed:
add system test dag with istio enabled

Add a new system test dag with istio enabled

Owen-CH-Leung committed:
add system test dag with istio enabled

Add a new system test dag with istio enabled

Owen-CH-Leung committed:
Merge branch 'add-istio-support-to-k8sOperator' of https://github.com/Owen-CH-Leung/airflow into add-istio-support-to-k8sOperator

Owen-CH-Leung committed:
Merge branch 'main' into add-istio-support-to-k8sOperator

Owen-CH-Leung committed:
Merge branch 'main' into add-istio-support-to-k8sOperator

Owen-CH-Leung committed:
Merge branch 'apache:main' into add-istio-support-to-k8sOperator

hussein-awala commented on airflow/providers/cncf/kubernetes/utils/pod_manager.py:
```diff
@@ -153,6 +153,31 @@ def container_is_terminated(pod: V1Pod, container_name: str) -> bool:
+        return False
+    return container_status.state.terminated.exit_code == 0
+
+
```
I think there is a problem with the rebase/merge, because these methods exist already in main branch

jtele2 replied:
I thought they had been merged already, just waiting on the new release so I don't have to build from source to use this. But still very eagerly waiting.

potiuk replied:
Well. You have conflict now and you are essentially 271 commits behind. I suggest you rebase rather than merge and solve conflicts you have now - it's not clear whether your merge has been done well. When you rebase you will see only your changes on top of the current main branch.

And yes, for changes like that, it can happen that it will take long time to merge them, and that you will have to resolve conflicts and rebase few times, and that you might have conflicts with other changes.

I strongly suggest to rebase it (and solve conflicts).

Owen-CH-Leung replied:
ahh... Sorry was too focused on making the code change & I performed merge instead of rebase. Will do rebase next time

hussein-awala commented on airflow/providers/cncf/kubernetes/operators/pod.py:
```diff
@@ -238,6 +238,7 @@ class KubernetesPodOperator(BaseOperator):
         state, or the execution is interrupted. If True (default), delete the
         pod; if False, leave the pod.
         Deprecated - use `on_finish_action` instead.
+    :param istio_enabled: Whether istio is enabled in k8s cluster. False by default.
```
I wonder if we can avoid adding a new argument, by adding a new option in `OnFinishAction`, wdyt?

Owen-CH-Leung replied:
Yes definitely. I've removed the extra param and added a new option in OnFinishAction. The logic for `await_pod_completion` & `process_pod_deletion` is updated also.

eladkal committed:
Prepare docs for RC3 providers (google) (#33253)

Usiel committed:
Enable pools to consider deferred tasks (#32709)

* Makes pools respect deferrable tasks (with extra setting)

See https://github.com/apache/airflow/discussions/21243

This commit makes pools consider deferred tasks if the `include_deferred` flag is set. By default a pool will not consider deferred tasks as occupied slots, but still show the number of deferred tasks in its stats.

---------

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>

Owen-CH-Leung committed:
Remove duplicated function

Owen-CH-Leung committed:
Remove duplicated function

Owen-CH-Leung committed:
Merge branch 'main' into add-istio-support-to-k8sOperator

bbovenzi committed:
Gantt chart: Use earliest/oldest ti dates if different than dag run start/end (#33215)

Owen-CH-Leung committed:
Add new enum in OnFinishAction, and revise test

Owen-CH-Leung committed:
Add new enum in OnFinishAction, and revise test

eumiro committed:
Refactor: Simplify code in providers/amazon (#33222)

Owen-CH-Leung committed:
Modify await_pod_completion logic

Owen-CH-Leung committed:
Modify await_pod_completion logic

potiuk committed:
Fix missing related PRs when cherry-picking (#33261)

The `airflow-github` tooling had a check if PR has been already
merged, but the PR did not work well when there was already a PR
cherry-picked that referred the PR in question in their log message.

The check was done via `git log --grep=#PR` essentially and so
it happenes that `--grep` matches not only the first line of
the message but the whole log message. So if the log message
contained reference to the cherry-picked PR it was seen as merged.

This change adds ( ) around and $ to the regexp to match not
only the PR number but also parentheses around.

Owen-CH-Leung committed:
Fix failing helm test

Owen-CH-Leung committed:
Merge branch 'add-istio-support-to-k8sOperator' of https://github.com/Owen-CH-Leung/airflow into add-istio-support-to-k8sOperator
